### ISSUE TITLE
catching another when:once event while job is running causes job to exit

### ISSUE THREAD

**Issue by tgross** [joyent/containerpilot#417]
Discovered this in https://github.com/autopilotpattern/mysql/pull/85. If a job has a `{ when: once: }` configuration but the source of that `once` can run more than once, then if the job is still running when the source fires again it will force the job to exit and fail.

The intended and tested to work workflow is as follows:

- `Job.Run` starts
- source event fires
- job catches event:
  - "starts remaining" is decremented
  - the job's `exec` is started
  - the job's `exec` finishes up and exits
- source event fires again
- job catches event:
  - sees "starts remaining" == 0
  - the job closes up

But we're missing the case of:

- `Job.Run` starts
- source event fires
- job catches event:
  - "starts remaining" is decremented
  - the job's `exec` is started
  - the job's `exec` is _still running_
- source event fires again
- job catches event:
  - sees "starts remaining" == 0
  - the job closes up, which kills the job's `exec`

---

An example configuration and its logs for this are as follows:

```json5
{
  consul: "localhost:8500",
  logging: { level: "DEBUG" },
  jobs: [
    {
      name: "preStart",
      exec: "sleep 20",
      when: {
        source: "consul-agent",
        once: "healthy"
      }
    },
    {
      name: "consul-agent",
      restarts: "unlimited",
      exec: "sleep 1000",
      health: {
        exec: "sleep 1",
        interval: 5,
        ttl: 10
      }
    }
  ],
  control: {
    socket: "./containerpilot.sock"
  }
}
```

```
2017/06/26 16:56:41 control: initialized router for control server
2017/06/26 16:56:41 event: {Startup global}
2017/06/26 16:56:41 consul-agent.Run start
2017/06/26 16:56:41 control: serving at ./containerpilot.sock
2017/06/26 16:56:46 check.consul-agent.Run start
2017/06/26 16:56:47 check.consul-agent exited without error
2017/06/26 16:56:47 event: {ExitSuccess check.consul-agent}
2017/06/26 16:56:47 check.consul-agent.Run end
2017/06/26 16:56:47 check.consul-agent.kill
2017/06/26 16:56:47 killing command 'check.consul-agent' at pid: 26067
2017/06/26 16:56:47 event: {StatusHealthy consul-agent}
2017/06/26 16:56:47 preStart.Run start
2017/06/26 16:56:51 check.consul-agent.Run start
2017/06/26 16:56:52 check.consul-agent exited without error
2017/06/26 16:56:52 event: {ExitSuccess check.consul-agent}
2017/06/26 16:56:52 check.consul-agent.Run end
2017/06/26 16:56:52 check.consul-agent.kill
2017/06/26 16:56:52 killing command 'check.consul-agent' at pid: 26105
2017/06/26 16:56:52 event: {StatusHealthy consul-agent}
2017/06/26 16:56:52 event: {Stopping preStart}
2017/06/26 16:56:52 event: {Stopped preStart}
2017/06/26 16:56:52 preStart.kill
2017/06/26 16:56:52 killing command 'preStart' at pid: 26068
2017/06/26 16:56:52 preStart exited with error: preStart: signal: killed
2017/06/26 16:56:52 event: {ExitFailed preStart}
2017/06/26 16:56:52 event: {Error preStart: signal: killed}
2017/06/26 16:56:52 preStart.Run end
```

---

I think the best way to handle this is that when we decrement the remaining starts in [`Job.processEvent`](https://github.com/joyent/containerpilot/blob/84493230dbe8db630973248e02308f51679460da/jobs/jobs.go#L286) and the count is now zero, that we swap out the `Job.startEvent` for an `events.NonEvent` so that we ignore it forever after.

**Comment by tgross** [comment]
> I think the best way to handle this is that when we decrement the remaining starts in `Job.processEvent` and the count is now zero, that we swap out the `Job.startEvent` for an `events.NonEvent` so that we ignore it forever after.

I've smoke-tested this idea and it seems to work at first glance. I'll build a test for it and I'll have a PR up for it tomorrow morning.

**Comment by tgross** [comment]
Fix should be in https://github.com/joyent/containerpilot/pull/418

**Comment by tgross** [comment]
Fix released in https://github.com/joyent/containerpilot/releases/tag/3.1.1

**Comment by eliasbrange** [comment]
This does not seem fixed to me when using v3.1.1.
```
# containerpilot --version
Version: '3.1.1'
GitHash: 'bbe6bdd'
```

Config:

```{
  consul: "localhost:8500",
  logging: {
    level: "DEBUG",
    format: "default",
    output: "stdout"
  },
  jobs: [
    {
      name: "consul-agent",
      exec: "sleep 1000",
      restarts: "unlimited",
      health: {
        exec: "sleep 1",
        interval: 5,
        ttl: 10
      }
    },
    {
      name: "consul-template",
      exec: "sleep 20",
      restarts: "unlimited",
      when: {
        source: "consul-agent",
        once: "healthy"
      }
    },
  ]
}```

Logs:

```2017/07/13 09:51:35 control: initialized router for control server                    
2017/07/13 09:51:35 event: {Startup global}                                           
2017/07/13 09:51:35 control: serving at /var/run/containerpilot.socket                
2017/07/13 09:51:35 consul-agent.Run start                                            
2017/07/13 09:51:40 check.consul-agent.Run start                                      
2017/07/13 09:51:41 check.consul-agent exited without error                           
2017/07/13 09:51:41 event: {ExitSuccess check.consul-agent}                           
2017/07/13 09:51:41 check.consul-agent.Run end                                        
2017/07/13 09:51:41 check.consul-agent.kill                                           
2017/07/13 09:51:41 killing command 'check.consul-agent' at pid: 14                   
2017/07/13 09:51:41 event: {StatusHealthy consul-agent}                               
2017/07/13 09:51:41 consul-template.Run start                                         
2017/07/13 09:51:45 check.consul-agent.Run start                                      
2017/07/13 09:51:46 check.consul-agent exited without error                           
2017/07/13 09:51:46 event: {ExitSuccess check.consul-agent}                           
2017/07/13 09:51:46 check.consul-agent.Run end                                        
2017/07/13 09:51:46 check.consul-agent.kill                                           
2017/07/13 09:51:46 killing command 'check.consul-agent' at pid: 18                   
2017/07/13 09:51:46 event: {StatusHealthy consul-agent}                               
2017/07/13 09:51:46 event: {Stopping consul-template}                                 
2017/07/13 09:51:46 event: {Stopped consul-template}                                  
2017/07/13 09:51:46 consul-template.kill                                              
2017/07/13 09:51:46 killing command 'consul-template' at pid: 17                      
2017/07/13 09:51:46 consul-template exited with error: consul-template: signal: killed
2017/07/13 09:51:46 event: {ExitFailed consul-template}                               
2017/07/13 09:51:46 event: {Error consul-template: signal: killed}                    
2017/07/13 09:51:46 consul-template.Run end```

**Comment by tgross** [comment]
Thanks for the good repro config @eliasbrange. Let me take a look and see where the `restarts` field may be fouling the logic.

**Comment by eliasbrange** [comment]
@tgross.

From PR #418, the if clause `if job.startsRemain == 0 && job.restartsRemain == 0` should probably only be `if job.startsRemain == 0`. Otherwise, if there are `restarts` left, the next time the `startEvent` is found, the `processEvent` function will return true on line 292 which will cause the `exec` to stop.

If I understand the code correctly, that is.

**Comment by tgross** [comment]
The tricky bits of this logic are there mostly because we allow `when->each` and `restarts` together, as this was intended to allow for something to happen "each time event X occurs, but a maximum of Y times." In retrospect I'm wondering whether that was ever a good idea, or if when we did it, it should have been separate from the concept of `restarts` which implies "restart the exec if it exits".

I'm not sure if we should eliminate that branch of the conditional because of how we need to handle the `each` case, but maybe it needs to be an `||` instead of a `&&`. My next step will be to make sure we have a test case for all the permutations -- I thought we did but clearly we're missing one.
